### DISCUSSION TITLE
Is there a directive to have the output generator inject Comments (for types that support them)?

### DISCUSSION THREAD

**Discussion by eli-xciv** [apple/pkl#376] [Q&A] [answered]
I'm looking for a way to specify a comment in a pkl file that will make it into a yaml file generated wtih `pkl eval -f yaml my.pkl`

example:
`my.pkl`
```
// Would Like to specify a comment here?
my: String = "My Pkl file"
```

Wanted output with `-f yaml`
```
# Would Like to specify a comment here?
my: My Pkl file
``` 

Is there currently a way to do this? Thanks in advance

**Comment by bioball** [comment, marked as answer]
There is not.

One existing pattern here is to change the `output.text` to include comments:

```pkl
foo = "Hello there"

output {
  text = """
    # Some comment
    \(new YamlRenderer {}.renderDocument(module))
    """
}
```
Although, this doesn't work as well if you want to add a comment in the middle of a file.

**Reply by eli-xciv** [reply]
Hrrrrm, would it be worthwhile to put in a feature request to be able to add comments within a file (for those types that support it).

**Reply by holzensp** [reply]
There are still quite a few ways to do this. Take a look at the default definition of `output` [here](https://github.com/apple/pkl/blob/main/stdlib/base.pkl#L93-L107). You could easily imagine a `comment: String` field added there, and changing the definition of `text` to `comment + "\n" + renderer.renderDocument(value)`.

However, that's not a general solution, because formats like JSON don't have comments (and so you'd only introduce leading `\n`). Since it can't be format agnostic, you'd be setting the `renderer` to the expected format anyway, so then you might as well set `output.text` the way @bioball did above.

One thing that _could_ be generally applicable is to add a `prefix: String = ""` and `suffix: String = ""`, such that `text = prefix + renderer.renderDocument(value) + suffix`. If anything, my point is that it'll take a little more refinement to turn this into a proper feature request; maybe continue this Discussion.

**Reply by eli-xciv** [reply]
@holzensp Thanks! I'll have to checkout using  fields and play around with the output renderer(s). 

At the start I was thinking of a more abstract implementation akin to a Java Annotation that would cause the renderer to inject a comment (if the backing file structure supports comments)

**Reply by holzensp** [reply]
If you want, you can. If you define `class MyAnnotation extends Annotation { ... }`, you can put a `@MyAnnotation` over object members. `pkl:reflect` will see those annotations. It does seem a somewhat roundabout way to achieve something that _sounds_ like it should be simpler (could this be an instance of the [XY problem](https://en.wikipedia.org/wiki/XY_problem)?).

**Reply by jamesrwu** [reply]
To add to this thread, I would love to have support for the use case when generating YAML config files, having a comment at the top of the file notifying others that this the YAML file was generated and should not be edited directly.

**Reply by StefMa** [reply]
> To add to this thread, I would love to have support for the use case when generating YAML config files, having a comment at the top of the file notifying others that this the YAML file was generated and should not be edited directly.

This is already possible.
See for example this one:
https://github.com/StefMa/pkl-gha/blob/238bb8b9e1924d4daabdb25c0777c8b6612cd6a8/Workflow.pkl#L689

This discussion is about having comments "at various places". Which is not possible atm.
But at the top or the bottom works just fine when you customize the output...
